spark_log_parser/loaders/__init__.py

```diff
@@ -61,58 +71,110 @@ def __iter__(self):
-                    break
+                self._buffer += chunk
 
-            # Trim off any data that is past the provided `size` and hold on to it for subsequent `read` calls
-            if len(content) > size:
-                self._trailing_data = content[size:]
-                content = content[:size]
+            self._consumed = True
+            return self._buffer
 
-        return content
+        else:
+            # Check if the data left in buffer if enough to satisfy the full read size...
+            if self._num_bytes_left_in_buffer() < size:
+                # If not, chop off everything that we've read out of the buffer, and start re-filling it. Reading
+                #  from our memoryview is actually slower here, since we end up having to convert that to bytes
+                #  anyway for our concatenation down below
+                self._buffer = self._buffer[self._buffer_position:]
+                self._buffer_position = 0
+
+                # iterate chunks until we've read at least `size` amount of data
+                for chunk in self._chunks:
+                    self._add_chunk_to_size(chunk)
+                    self._buffer += chunk
+                    if len(self._buffer) >= size:
+                        break
+
+                # Freeze this buffer for subsequent reads from it. We use a memoryview in order to avoid making copies
+                #  of the data for every single read issued.
+                self._buffer_memview = memoryview(self._buffer).cast("B").toreadonly()
+
+            # If we have enough data in the buffer to satisfy this read, return that from the buffer and adjust
+            #  our read position accordingly
+            if self._num_bytes_left_in_buffer() >= size:
+                start_pos = self._buffer_position
+                self._buffer_position += size
+                return self._buffer_memview[start_pos:self._buffer_position]
+            else:
+                # Otherwise, we don't have enough data, which means that this is the last read for this iterator.
+                #  So now just return a copy of the data remaining in the buffer and mark this as consumed to prevent
+                #  future reads
+                content = self._buffer_memview[self._buffer_position:]
+                self._consumed = True
+                return content
+
+    _LINE_CHUNK_SIZE = 1024 * 1024
 
     def iter_lines(self) -> Iterator[bytes]:
         """
         Returns a stream of lines from self._chunks.
         """
-        for chunk in self._chunks:
-            self._add_chunk_to_size(chunk)
-            if self._trailing_data:
-                chunk = self._trailing_data + chunk
-
-            lines = chunk.splitlines()
+        trailing_data = bytearray()
+        # We use a relatively large chunk_size here to minimize the number of times we have to enter this loop/copy data
+        while chunk := self.read(self._LINE_CHUNK_SIZE):
```

Comment:
For some instances of `self._chunks`, `for chunks in self._chunks` is already returning single-lines to us. Not that we have a memory-efficient `self.read` implementation, we can just use that so that we can process data more than one-line-at-a-time in this function.